coredns/plugin/zone.go

```diff
@@ -0,0 +1,233 @@
+		}
+	}
+
+	if ep.RecordType == endpoint.RecordTypeA {
+		for _, t := range ep.Targets {
+			a := &dns.A{Hdr: dns.RR_Header{Name: dns.Fqdn(ep.DNSName), Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: uint32(ep.RecordTTL)},
+				A: net.ParseIP(t)}
+			rrs = append(rrs, a)
+		}
+	}
+
+	if ep.RecordType == endpoint.RecordTypeAAAA {
+		for _, t := range ep.Targets {
+			aaaa := &dns.AAAA{Hdr: dns.RR_Header{Name: dns.Fqdn(ep.DNSName), Rrtype: dns.TypeAAAA, Class: dns.ClassINET, Ttl: uint32(ep.RecordTTL)},
+				AAAA: net.ParseIP(t)}
+			rrs = append(rrs, aaaa)
+		}
+	}
+
+	if ep.RecordType == endpoint.RecordTypeTXT {
+		txt := &dns.TXT{Hdr: dns.RR_Header{Name: dns.Fqdn(ep.DNSName), Rrtype: dns.TypeTXT, Class: dns.ClassINET, Ttl: uint32(ep.RecordTTL)},
+			Txt: ep.Targets}
+		rrs = append(rrs, txt)
+	}
+
+	if ep.RecordType == endpoint.RecordTypeCNAME {
+		cname := &dns.CNAME{Hdr: dns.RR_Header{Name: dns.Fqdn(ep.DNSName), Rrtype: dns.TypeCNAME, Class: dns.ClassINET, Ttl: uint32(ep.RecordTTL)},
+			Target: dns.Fqdn(ep.Targets[0])}
+		rrs = append(rrs, cname)
+	}
+
+	for i := range rrs {
+		rrd := rrData{}
+		if wProp, wExists := ep.GetProviderSpecificProperty(v1alpha1.ProviderSpecificWeight); wExists {
+			weight, err := strconv.ParseInt(wProp, 10, 64)
+			if err != nil {
+				weight = 0
+			}
+			rrd.weight = &weight
+		} else if gProp, gExists := ep.GetProviderSpecificProperty(v1alpha1.ProviderSpecificGeoCode); gExists {
+			rrd.geo = &gProp
+		}
+		z.rrData[rrs[i]] = rrd
+		z.file.Insert(rrs[i])
+	}
+
+	return nil
+}
+
+func (z *Zone) Lookup(ctx context.Context, state request.Request, qname string) ([]dns.RR, []dns.RR, []dns.RR, file.Result) {
+	return z.file.Lookup(ctx, state, qname)
+}
+
+func (z *Zone) RefreshFrom(newZ *Zone) {
+	// copy elements we need
+	z.file.Lock()
+	z.file.Apex = newZ.file.Apex
+	z.file.Tree = newZ.file.Tree
+	z.rrData = newZ.rrData
+	z.file.Unlock()
+}
+
+// parseWeightedAnswers takes a slice of answers for a dns name and reduces it down to a single answer based on weight.
+func (z *Zone) parseWeightedAnswers(_ context.Context, _ request.Request, wrrs []dns.RR) []dns.RR {
+	log.Debugf("parsing weighted answers for %s", wrrs[0].Header().Name)
+	var answer *dns.RR
+	var weightedRRs []weightedRR
+
+	for _, r := range wrrs {
+		if w := z.rrData[r].weight; w != nil {
+			weightedRRs = append(weightedRRs, weightedRR{r, *w})
+		}
+	}
+
+	if weightedRRs != nil {
+		wRRSet := newWeightedRRSet(weightedRRs)
+		answer = wRRSet.getTopRR()
+	}
+
+	if answer == nil {
+		answer = &wrrs[0]
+	}
+
+	return []dns.RR{*answer}
+}
+
+// parseGeoAnswers takes a slice of answers for a dns name and reduces it down to a single answer based on geo.
+func (z *Zone) parseGeoAnswers(ctx context.Context, request request.Request, grrs []dns.RR) []dns.RR {
+	log.Debugf("parsing geo answers for %s", grrs[0].Header().Name)
+	log.Debugf("source ip is %s", request.IP())
+	var answer *dns.RR
+	var geoRRs []geodRR
+
+	roundRobinShuffle(grrs)
```

Comment:
This should probably be removed, but it makes me wonder what we would like the default behaviour to be when we can't match any geo to the current request. Do we want it to always just return the first RR in the list?

Currently this will cause it to randomise what it would return if there are multiple geos available.

Reply:
I think we should leverage the default GEO like AWS

Reply:
Added a follow on task to look into this https://github.com/Kuadrant/dns-operator/issues/409